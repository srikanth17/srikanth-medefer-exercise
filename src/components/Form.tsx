import React from 'react';
import { Field, InjectedFormProps, reduxForm } from 'redux-form';
import FormField from './FormField';
import formFields from '../utils/formFields';

interface FormProps extends InjectedFormProps {
  onFormSubmit: (values: any) => void;
}

const Form = (props: FormProps) => {
  const { handleSubmit, error, onFormSubmit } = props;

  return (
    <div>
      <form onSubmit={handleSubmit(onFormSubmit)}>
        {formFields.map(field => (
          <Field
            key={field.name}
            label={field.label}
            type="text"
            name={field.name}
            component={FormField}
          />
        ))}
        {error && <strong>{error}</strong>}
        <button type="submit" className="green btn-flat right white-text">
          Submit
          <i className="material-icons right">done</i>
        </button>
      </form>
    </div>
  );
};

export default reduxForm<{}, FormProps>({
  form: 'Form',
  destroyOnUnmount: false,
})(Form);
