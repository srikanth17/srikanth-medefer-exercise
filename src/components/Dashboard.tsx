import React, { useState } from 'react';
import { SubmissionError } from 'redux-form';
import validateEmail from '../utils/validateEmail';
import Form from './Form';
import Summary from './Summary';
import { FormData } from '../types/types';

const Dashboard = () => {
  const [showForm, setShowForm] = useState(false);
  const [formSubmitted, setFormSubmitted] = useState(false);

  const submit = (values: FormData) => {
    if (!values.firstName) {
      throw new SubmissionError({
        firstName: 'The First name field is required',
        _error: 'Form cannot be submitted!',
      });
    } else if (!values.surname) {
      throw new SubmissionError({
        surname: 'The Surname field is required',
        _error: 'Form cannot be submitted!',
      });
    } else if (!validateEmail(values.email || '')) {
      console.log(validateEmail(values.email || ''));
      throw new SubmissionError({
        email: 'The email is invalid or the field is empty',
        _error: 'Form cannot be submitted!',
      });
    } else {
      setFormSubmitted(true);
    }
  };

  if (formSubmitted) {
    return <Summary />;
  } else if (showForm) {
    // @ts-ignore
    return <Form onFormSubmit={submit} />;
  } else {
    return (
      <div>
        <h5>Please click the ADD button to fill the form</h5>
        <button className="btn-large red" onClick={() => setShowForm(true)}>
          Add<i className="material-icons right">add</i>
        </button>
      </div>
    );
  }
};

export default Dashboard;
