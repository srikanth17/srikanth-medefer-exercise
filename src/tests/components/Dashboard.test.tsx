import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import Dashboard from '../../components/Dashboard';
import Form from '../../components/Form';

let wrapper: ShallowWrapper;

beforeEach(() => {
  wrapper = shallow(<Dashboard />);
});

test('should render Dashboard correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should render ADD button in Summary component', () => {
  expect(wrapper.find('button').exists()).toBe(true);
});
