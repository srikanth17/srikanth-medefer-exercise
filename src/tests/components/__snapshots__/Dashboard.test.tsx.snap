// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Dashboard correctly 1`] = `
<div>
  <h5>
    Please click the ADD button to fill the form
  </h5>
  <button
    className="btn-large red"
    onClick={[Function]}
  >
    Add
    <i
      className="material-icons right"
    >
      add
    </i>
  </button>
</div>
`;
